src/Storages/StorageMerge.cpp

```diff
@@ -893,19 +951,22 @@ void ReadFromMerge::convertingSourceStream(
         });
     }
 
-    {
-        auto convert_actions_dag = ActionsDAG::makeConvertingActions(builder.getHeader().getColumnsWithTypeAndName(),
-                                                                     header.getColumnsWithTypeAndName(),
-                                                                     ActionsDAG::MatchColumnsMode::Name);
-        auto actions = std::make_shared<ExpressionActions>(
-            convert_actions_dag,
-            ExpressionActionsSettings::fromContext(local_context, CompileExpressions::yes));
+    ActionsDAG::MatchColumnsMode convert_actions_match_columns_mode = ActionsDAG::MatchColumnsMode::Name;
 
-        builder.addSimpleTransform([&](const Block & stream_header)
-        {
-            return std::make_shared<ExpressionTransform>(stream_header, actions);
-        });
-    }
+    if (local_context->getSettingsRef().allow_experimental_analyzer && processed_stage != QueryProcessingStage::FetchColumns)
+        convert_actions_match_columns_mode = ActionsDAG::MatchColumnsMode::Position;
```

Comment:
@kitaisreal can you please clarify why analyzer needs matching by positions? doesn't it can cause troubles?

Reply:
It needs this because names in case of storage Merge are not reliable in new infrastructure, but positions are. Storage merge is still not fully supported in new infrastructure https://github.com/ClickHouse/ClickHouse/issues/42648, maybe some of the broken tests are related to this change.

Reply:
Yeah, I have the same thoughts, that somewhere the names had been changed due to analyzer appends index, but tried a simple case with matching by names and it worked. So I thought that maybe you remember the particular case.

Also one more question, why does analyzer appends the index to the column name?https://github.com/ClickHouse/ClickHouse/blob/8d0a201468b5408aab8a072628f321a9ceb1b0bb/src/Planner/PlannerContext.cpp#L27

Reply:
We make all column names unique to properly support JOINs, when columns from different tables have the same name. We rely on unique column names during query plan build and query plan optimizations. 
If you have any additional questions related to analyzer, lets discuss it in main migration ticket, it seems that no one will find answers in this old ticket.